keyboards/splitography/keymaps/beaklmu/common/keycode_functions.h

```diff
@@ -0,0 +1,660 @@
+static uint8_t thumb = 0;
+
+// LEFT (KC_SPC, S(KC_BSLS)), RIGHT (KC_LEFT, S(KC_LEFT)) opposite thumb combinations, see process_record_user()
+// up,   up   -> _BASE
+// up,   down -> _SYMBOL
+// down, up   -> _NUMBER
+// down, down -> _MOUSE                     // see layer keycodes that raise mouse layer
+#define THUMBS_DOWN _MOUSE                  // layer
+
+static uint8_t overlayer = 0;
+
+// left right thumb layer combinations
+void thumb_layer(keyrecord_t *record, uint8_t side, uint8_t shift, uint16_t keycode, uint8_t thumb_dn_layer, uint8_t thumb_up_layer)
+{
+  if (record->event.pressed) {
+    // layer_on via tap_layer(), see process_record_user()
+    key_timer = timer_read();
+    thumb     = thumb | side;
+  }
+  else {
+    layer_off(thumb_dn_layer);
+    // opposite thumb_layer() thumb may have switched effective layer!
+    if (overlayer) {
+      layer_off(overlayer);
+      overlayer = 0;
+    }
+    if (!key_press(shift, keycode)) {
+      layer_off(THUMBS_DOWN);               // both thumbs needed
+      // opposite thumb down? see left right combination layer table above
+      if (thumb & (side == LEFT ? RIGHT : LEFT)) {
+        layer_on(thumb_up_layer);
+        overlayer = thumb_up_layer;
+      }
+    }
+    clear_mods();
+    thumb     = thumb & ~side;
+    key_timer = 0;
+  }
+}
+
+// #ifdef STENO_ENABLE
+// // LT for steno keycode
+// void stn_layer(keyrecord_t *record, uint16_t keycode, uint8_t layer)
+// {
+//   if (record->event.pressed) {
+//     key_timer = timer_read();
+//     if (keycode) {
+//       process_steno(keycode, record);
+//     }
+//     layer_on(layer);
+//   }
+//   else {
+//     layer_off(layer);
+//     if (keycode) {
+//       if (timer_elapsed(key_timer) < TAPPING_TERM) {
+//         process_steno(keycode, record);
+//       }
+//       else {
+//         // clear pressed state (request push of updated) process_steno.c and .h
+//         // steno_clear_state();
+//       }
+//     }
+//     key_timer = 0;
+//   }
+// }
+// #endif
+
+// LT for S(keycode)
+void lt_shift(keyrecord_t *record, uint16_t keycode, uint8_t layer)
+{
+  if (record->event.pressed) {
+    key_timer = timer_read();
+    layer_on(layer);
+  }
+  else {
+    layer_off(layer);
+    // for shifted keycodes, hence, LT_SHIFT
+    key_press(SHIFT, keycode);
+    clear_mods();
+    key_timer = 0;
+  }
+}
+
+// set layer asap to overcome macro latency errors, notably tap dance and LT usage
+// this routine inexplicably (?) sets layer_on() faster than can be done in thumb_layer()
+void tap_layer(keyrecord_t *record, uint8_t layer)
+{
+  if (record->event.pressed) {
+    layer_on(layer);
+  }
+  else {
+    layer_off(layer);
+  }
+}
+
+// ..................................................................... Keymaps
+
+// void persistant_default_layer_set(uint16_t default_layer)
+// {
+//   eeconfig_update_default_layer(default_layer);
+//   default_layer_set            (default_layer);
+// }
+
+void clear_layers(void)
```

Comment:
`layer_move(0)` does the same thing.

Reply:
Thank you for your very helpful responses. I hope to have time to refactor this and resubmit in a different pull request. I will close this now, but I thank you for your detailed response. I appreciate your time and help.